### PR TITLE
[Feat][Ops] implement Argmax/Argmin dim=None full-tensor reduction

### LINKED ISSUE #1090: [Feat][Ops] implement Argmax/Argmin dim=None full-tensor reduction
### Symptom / Motivation

Discovered during PR #1084 ([Refactor][Ops] align 7 reduction op __init__ signatures with manifest). `ArgmaxFwdOp` and `ArgminFwdOp` were promoted to `status: implemented`, but reviewer (Ibuki-wind) flagged that the manifest declares `dim: int | None` (matching `torch.argmax(x)` / `torch.argmin(x)` full-tensor reduction semantics) while the implementation's `_validate_dim()` rejects `None`. Promotion was reverted to `spec-only` in commit 3be5d00 with an inline gap marker; this issue tracks closing the gap.

When `dim is None`, PyTorch flattens the input and returns a scalar `int64` index into the flattened tensor (or a fully reduced shape when `keepdim=True`).

### Root Cause Analysis

`_validate_dim()` was authored before the argreduce manifest entries were aligned to PyTorch in PR #1072. Validator L1+ checks pass because the signature type is correct in the manifest; semantic gap (None being rejected at runtime) surfaces only on reviewer audit.

### Related Files

- `tileops/ops/reduction/argmax.py`
- `tileops/ops/reduction/argmin.py`
- `tileops/ops/reduction/reduce.py` (or wherever `_validate_dim` lives)
- `tileops/kernels/reduction/argreduce.py` (kernel may need a flat-reduction entry)
- `tests/ops/test_argreduce.py`
- `tileops/ops_manifest.yaml` (entries `ArgmaxFwdOp`, `ArgminFwdOp`)

## Goal

Implement `dim=None` full-tensor reduction for `ArgmaxFwdOp`/`ArgminFwdOp` matching `torch.argmax`/`torch.argmin` semantics, and re-flip both to `status: implemented`.

## Plan

1. Allow `None` in `_validate_dim()`; when `None`, treat as a contiguous flatten + 1D reduction.
2. Add a flat-reduction kernel path in `ArgreduceKernel` (or reuse the existing 1D specialization with reshape-to-flat at the op layer).
3. Output shape: `()` when `keepdim=False`; all-ones when `keepdim=True`. Index value is into the flattened tensor.
4. Add tests covering `dim=None` × `keepdim={True, False}` for fp16/bf16/fp32 inputs and at least 2D and 3D shapes.
5. Run `scripts/validate_manifest.py --check-op ArgmaxFwdOp` and `--check-op ArgminFwdOp`.
6. Flip both manifest entries to `status: implemented`; remove the inline gap-marker comments.

## Constraints

- Must not regress PR #1084.
- Index dtype remains `int64` (manifest fixed). Index semantics must match PyTorch (flattened-tensor index when `dim is None`).
- Manifest stays untouched (source of truth).

## Acceptance Criteria

- [ ] AC-1: `ArgmaxFwdOp(dim=None)(x)` and `ArgminFwdOp(dim=None)(x)` match `torch.argmax(x)`/`torch.argmin(x)` for `x.ndim ∈ {1, 2, 3, 4}` and dtypes in {fp16, bf16, fp32}; `keepdim={True, False}` paths agree on shape and value.
- [ ] AC-2: New tests added under `tests/ops/test_argreduce.py` covering `dim=None` cases pass.
- [ ] AC-3: `scripts/validate_manifest.py --check-op ArgmaxFwdOp` and `--check-op ArgminFwdOp` report "All manifest checks passed".
- [ ] AC-4: Both manifest entries show `status: implemented` with the `dim=None` inline gap-marker comment removed.

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1090

## Summary

- Allow `dim=None` in `ArgmaxFwdOp`/`ArgminFwdOp` to perform full-tensor argreduce, matching `torch.argmax(x)` / `torch.argmin(x)` semantics (flattened-tensor index, int64).
- `_validate_dim` accepts `None`; the op layer flattens the input and dispatches the existing 1D argreduce path. Output is `()` when `keepdim=False` and all-ones when `keepdim=True`.
- Align constructor defaults to the manifest: `dim: Optional[int] = None` (was `dim: int = -1`) for both ops, so `ArgmaxFwdOp()(x)` matches `torch.argmax(x)` instead of silently doing last-axis.
- Flip `ArgmaxFwdOp` / `ArgminFwdOp` manifest entries to `status: implemented` and remove the inline `dim=None` gap-marker comments (AC-4).
- Drive `dim` through manifest workload entries: declared `dim: -1` on the four Argmax/Argmin workloads and switched `bench_argreduce.py` to `workloads_to_params(_OP, include_extra=True)`, instantiating the op + baseline from `extra["dim"]` so the benchmark is fully manifest-driven.

## Test plan

- [x] AC-1: `ArgmaxFwdOp(dim=None)(x)` / `ArgminFwdOp(dim=None)(x)` match `torch.argmax(x)` / `torch.argmin(x)` across dtype dispatch (fp16/bf16/fp32) and ndim shape branches (1D / 3D / 4D), with `keepdim ∈ {False, True}` covered in each case.
- [x] AC-2: New `dim=None` cases in `tests/ops/test_argreduce.py` pass (`pytest tests/ops/test_argreduce.py -m smoke -q` → 64 passed).
- [x] AC-3: `python scripts/validate_manifest.py --check-op ArgmaxFwdOp` and `--check-op ArgminFwdOp` both report "All manifest checks passed".
- [x] AC-4: Both manifest entries show `status: implemented`; inline `dim=None` gap-marker comments removed.
- [x] Bench collection (`pytest --collect-only benchmarks/ops/bench_argreduce.py`) → 8 cases, `extra={"dim": -1}` for each.
- [x] pre-commit passed
- [x] pytest passed

## Structural Readiness

- **Correctness & Safety**: PASS — `Op.forward` rejects non-CUDA tensors, dtype mismatch, and 0D inputs before kernel launch; `dim=None` derives `M=1`, `N=numel` via `flatten_for_multidim`; runtime validation uses `ValueError` / `IndexError`; output dtype is `int64`.
- **Kernel Structure**: SKIP — no kernel code changed for this issue; existing `ArgreduceKernel` already uses `T.prim_func` inside `tilelang.jit` and the builder is `lru_cache(maxsize=32)`.
- **Op Structure**: PASS — Op pre/post-processing remains in `_ReduceOpBase`; `ArgreduceKernel` retains `torch.library.custom_op` and `register_fake`; `default_config` / `autotune_configs` exist. Constructor defaults now match the manifest signature exactly.
- **Benchmark**: PASS — bench is driven by manifest workload params (`include_extra=True`); no hard-coded op params remain in `bench_argreduce.py`.
- **Delivery**: PASS — targeted unit tests pass; `dim=None` tests compare against `torch.argmax` / `torch.argmin`.

## Test node delta

```
File                           Base    HEAD    Delta
----------------------------------------------------
tests/ops/test_argreduce.py      70      84      +14
----------------------------------------------------
TOTAL                            70      84      +14
```

**Justification:** The dim=None coverage follows testing.md §Test case policy / testing-budget §11–12 — dtype and ndim are not crossed unless they guard distinct code paths. The fixture has 6 smoke rows (dtype dispatch on a single 2D shape × {fp16, bf16, fp32}, plus ndim shape coverage 1D / 3D / 4D in fp16 only since the flatten path is dtype-independent) plus 1 full row for non-aligned tail handling, consumed by two test functions (argmax / argmin) that each fold `keepdim ∈ {False, True}` into the body. Net: +14 nodes for the new feature.

### CODE REVIEW

File: tests/ops/test_argreduce.py [hunk 1]
```diff
@@ -5,12 +5,25 @@
 Uses exact match (torch.equal) instead of allclose.
 """
 
+from typing import cast
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py:8]: Reviewer determined this is a false positive:

active Ruff config does not require annotations (ANN is disabled), and the suggested import-only change would introduce an unused Any import under Pyflakes F401 unless paired with an annotation. The existing helper remains valid.

Current HEAD: b563c35

File: tests/ops/test_argreduce.py [hunk 2]
```diff
@@ -5,12 +5,25 @@
...
+
 import pytest
 import torch
 
 from tests.test_base import FixtureBase, TestBase
 from workloads.argreduce import ArgmaxTest as _ArgmaxWorkload
 
+
+def _call(op, x: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py:17]: Addressed in b563c35.

Evidence:
- Active Ruff config disables ANN (annotation) checks in `pyproject.toml`, so missing annotations on the `_call` helper are not a lint failure.
- The suggested import-only change (`from typing import Any, cast`) would introduce an unused `Any` import, triggering Pyflakes F401, unless paired with a signature annotation edit.
- This matches the analysis already applied to the sibling thread on line 8 (resolved). No behavior or required verification is affected.

Current HEAD: b563c35

File: benchmarks/ops/bench_argreduce.py [hunk 3]
```diff
@@ -26,7 +26,7 @@ def test_argmax_bench(shape: tuple, dtype: torch.dtype) -> None:
     workload = ArgmaxTest(shape, dtype)
     inputs = workload.gen_inputs()
 
-    op = ArgmaxFwdOp(dtype=dtype)
+    op = ArgmaxFwdOp(dtype=dtype, dim=-1)
```
Comment (Ibuki-wind) [inline benchmarks/ops/bench_argreduce.py]: `ArgmaxFwdOp(dtype=dtype, dim=-1)` hard-codes an op parameter that now disagrees with the manifest workload default (`dim` defaults to `null` after this PR), so the benchmark is no longer driven by the manifest source of truth -> declare `dim: -1` on the Argmax manifest workloads that are meant to reduce the last axis, call `workloads_to_params(_ARGMAX_OP, include_extra=True)`, and instantiate the op/baseline from those workload params instead of forcing `dim=-1` here.

File: benchmarks/ops/bench_argreduce.py [hunk 4]
```diff
@@ -59,7 +59,7 @@ def test_argmin_bench(shape: tuple, dtype: torch.dtype) -> None:
     workload = ArgminTest(shape, dtype)
     inputs = workload.gen_inputs()
 
-    op = ArgminFwdOp(dtype=dtype)
+    op = ArgminFwdOp(dtype=dtype, dim=-1)
```
Comment (Ibuki-wind) [inline benchmarks/ops/bench_argreduce.py]: `ArgminFwdOp(dtype=dtype, dim=-1)` hard-codes an op parameter that now disagrees with the manifest workload default (`dim` defaults to `null` after this PR), so the benchmark is no longer driven by the manifest source of truth -> declare `dim: -1` on the Argmin manifest workloads that are meant to reduce the last axis, call `workloads_to_params(_ARGMIN_OP, include_extra=True)`, and instantiate the op/baseline from those workload params instead of forcing `dim=-1` here.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Benchmark wiring needs to stay manifest-driven before these entries can be treated as ready.

### Advisory
- This PR spans manifest, implementation, tests, benchmarks, and workloads during the migration period; treating the trust-boundary crossing as advisory only.

File: benchmarks/ops/bench_argreduce.py (hunk 3 above)
Comment (lcy-seso) [reply benchmarks/ops/bench_argreduce.py]: Adopted. Declared `dim: -1` on the Argmax/Argmin manifest workload entries (`tileops/manifest/reduction.yaml`) and switched `bench_argreduce.py` to `workloads_to_params(_ARGMAX_OP, include_extra=True)`, instantiating the op as `ArgmaxFwdOp(dtype=dtype, **extra)` and the baseline as `x.argmax(dim=extra["dim"])` so the bench is fully manifest-driven. See b08af8d.

File: benchmarks/ops/bench_argreduce.py (hunk 4 above)
Comment (lcy-seso) [reply benchmarks/ops/bench_argreduce.py]: Adopted. Declared `dim: -1` on the Argmin manifest workload entries (`tileops/manifest/reduction.yaml`) and switched the Argmin bench to `workloads_to_params(_ARGMIN_OP, include_extra=True)`, instantiating `ArgminFwdOp(dtype=dtype, **extra)` and the baseline as `x.argmin(dim=extra["dim"])`. See b08af8d.

**Review by Ibuki-wind — DISMISSED** [review]
### Overall
Clean — manifest workload params now drive argreduce benchmark op and PyTorch baseline dimensions.

### Prior blockers
All 2 prior blockers resolved — see thread replies.

### Advisory
- This PR spans manifest, implementation, tests, benchmarks, and workloads during the migration period; advisory only.

**Review by Ibuki-wind — APPROVED** [review]
### Overall
Clean — the Round 3 test shrink preserves dtype, ndim, keepdim, and tail coverage without adding new risks.

### Prior blockers
All 2 prior blockers resolved — see thread replies.

### Advisory
- This PR spans manifest, implementation, tests, benchmarks, and workloads during the migration period; advisory only.
